ui/src/main/webapp/src/app/components/reports/tag-filter.service.ts

```diff
@@ -0,0 +1,46 @@
+import { ReportFilter, Tag } from "windup-services";
+import { TechnologyTagModel } from "../../generated/tsModels/TechnologyTagModel";
+
+export class TagFilterService {
+
+    constructor(private _reportFilter: ReportFilter) {
+    }
+
+    tagMatches(tag: TechnologyTagModel | string):boolean {
+        // Always match if there is no filter
+        if (!this._reportFilter)
+            return true;
+
+        // Always match null tags
+        if (!tag)
+            return true;
+
+        let tagValue = tag;
+        if (tagValue instanceof TechnologyTagModel)
+            tagValue = (<TechnologyTagModel>tagValue).name;
+
+        if (!this._reportFilter.includeTags)
+            return true;
+
+        if (!this._reportFilter.excludeTags)
+            return true;
+
+        let matches = false;
+        if (this.tagValueInCollection(this._reportFilter.excludeTags, tagValue)) {
+            matches = false;
+        }
+
+        if (!this._reportFilter.includeTags.length || this.tagValueInCollection(this._reportFilter.includeTags, tagValue))
+            matches = true;
+
+        return matches;
+    }
```

Comment:
I'm not sure if logic of this method is right.

What if I set only exclude tags and includeTags are empty? Then it will always return true. But I want to exclude that tag. Similar thing for other case where I have excludeTags empty and includeTags set.

IMHO it should work this way:
1) We should have some source with higher priority, I suggest includeTags.
2) If includeTags are empty, we only care about excludeTags. If tag is in excludeTags,  return false, otherwise return true (tag is not explicitly excluded)
3) if includeTags are not empty, look if tag is included in it. If it is, return true. Otherwise return false.

(at least I have it implemented this way on backend for migration issues).